Reduce RAM usage considerably on controllers

TD-er opened the pull request:
Loading of the Controller settings was done a lot and this controller settings struct is also quite large.
On fragmented memory it is possible this would lead to a crash.

For a lot of calls only a single string and/or bool was needed, so these can be cached in the controller itself.
Also the MQTT publish calls were making quite a lot of rather large copy calls when handling the strings for the topic and the message.

This also could lead to some very strange and hard to reproduce crashes.

TD-er committed:
[Controller] Memory optimization for sending SENSOR_TYPE_STRING

String was copied before sending to the MQTT publish.
Plus copied another time for the log if LOG_LEVEL_DEBUG was active.

TD-er committed:
[Cleanup] Try to use move constructor when adding to controller queue

Still to do: create a proper move operator for queue_element_single_value_base

TD-er committed:
[MQTT] More efficient MQTT copy incoming messages

TD-er committed:
[MQTT import] Check if MakeControllerSettings failed due to out of RAM

TD-er committed:
[Controller] Check if controller host/IP is set before MQTT connect

TD-er committed:
[MQTT] Add function to check MQTT queue full state

TD-er committed:
[Cleanup] Free ControllerSettings as soon as possible to free memory

Quote often the ControllerSettings are only used to get one or two values.
No need to keep this ~800 bytes large object allocated for more than absolutely needed.

TD-er committed:
[Controller] Cache needed controller info to reduce memory usage

TD-er committed:
[Controller] Check if MakeControllerSettings failed due to low RAM

Add checks to see if allocating the relative large ControllerSettingsStruct failed.
This can be caused by either running out of memory or when memory gets fragmented.

TD-er commented:
I created some test builds.

Would really appreciate if these could be tested with some of the most used controllers like any MQTT controller or HTTP controller.
Also the MQTT import should be tested.

- [ESPEasy_ESP32_mega-20200608-14-PR_3122.zip](https://www.dropbox.com/s/5wws9qrcfamna8q/ESPEasy_ESP32_mega-20200608-14-PR_3122.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200608-14-PR_3122.zip](https://www.dropbox.com/s/a1qb3ioii4a9vom/ESPEasy_ESP82xx_mega-20200608-14-PR_3122.zip?dl=0)

TD-er committed:
Fix merge issue WiFiConnected changed to NetworkConnected

ghtester commented:
Is there a way how could I selfcompile this PR using Vagrant and my Custom.h please?
I suppose the test build above won't have included some plugins I need so the testing is limited in my case.

TD-er commented:
The test build is always the set of PlatformIO environments that result in a valid build (e.g. not too large)
So all builds defined in the .ini files should be there.

But you can also build it in your PIO environment by running the build script in the tools dir with the parameter `-p 3122` for pull request 3122 (this PR)
Then it will checkout Git on that pull request.
After that's loaded, you can kill the script and activate the Python virtual environment:
`source ~/.venv/python3.8/bin/activate` 
Then just run the build via
`pio run -e custom_ESP8266_4M1M`

That one will build a version with the Custon.h that was already copied when you booted the Vagrant environment.

TD-er commented:
Oh and by the way.
If you are testing, the total amount of free RAM isn't reduced a lot by this PR.
But the needed amount of RAM during processing of MQTT publish and such will be less.
Tested here for a while in another setup that was constantly crashing, but with these fixes now running for days (compared to minutes)
Especially with highly fragmented memory and processing long MQTT messages (600 byte messages + topic), it was very unstable.

TD-er committed:
[P001 Switch] Reduce stack usage on PLUGIN_WEBFORM_LOAD

TD-er committed:
[Blynk] Reduce stack usage on Blynk_get

There is still room for improvement as we still try to create the entire string in a fixed size buffer.
Maybe we can also send out the data in chunks?
Or determine the needed array size at runtime.

TD-er committed:
[Web Frontend] Reduce stack usage on DST configuration

v-a-d-e-r commented:
I merged the controller and MQTT stuff yesterday on my test node and can not see any improvement, but also no new issue... ;-)
BTW: For what is the brace {...} in the heap good for? Example:
```c++
      { <--??
        addFormNumericBox(F("Dim value"), F("p001_dimvalue"), PCONFIG(1), 0, 255);
        String options[2];
        options[0] = F("Switch");
        options[1] = F("Dimmer");
        int optionValues[2]   = { PLUGIN_001_TYPE_SWITCH, PLUGIN_001_TYPE_DIMMER };
        const byte switchtype = P001_getSwitchType(event);
        addFormSelector(F("Switch Type"), F("p001_type"), 2, options, optionValues, switchtype);

        if (switchtype == PLUGIN_001_TYPE_DIMMER)
        {
          addFormNumericBox(F("Dim value"), F("p001_dimvalue"), PCONFIG(1), 0, 255);
        }
      } <--??
```

TD-er commented:
I added code block markers to your post for readability.

The { ... } mark a scope in the code.
Usually you only see it when doing an `if` or `while` or defining a function. 
But you can also have them in the code like this.
What it does is to make sure the variables which are declared on the stack will be destructed after they left the scope.
Especially the array variables are quite big (the ones declared with square brackets).
And if you allocate variables in a scope on the stack, they will all be active at the same time and thus you need to sum their size to see how much stack you use.
But now I grouped their use into a much smaller scope, so they are removed from the stack when you leave the scope.

A very simple but effective way to reduce the stack usage.

v-a-d-e-r commented:
Ah, ok. Thanks

TD-er committed:
Merge branch 'mega' into bugfix/memory_optimizatons
